src/finite_field.rs

```diff
@@ -92,95 +89,71 @@ impl std::ops::DivAssign for Field {
     }
 }
 
+impl PartialEq for Field {
+    fn eq(&self, rhs: &Self) -> bool {
+        SMALL_FP.from_elem(self.0) == SMALL_FP.from_elem(rhs.0)
+    }
+}
+
+impl Eq for Field {}
+
 impl Field {
     /// Modular exponentation
     pub fn pow(self, exp: Self) -> Self {
-        // repeated squaring
-        let mut base = self;
-        let mut exp = exp.0;
-        let mut result: Field = Field(1);
-        while exp > 0 {
-            while (exp & 1) == 0 {
-                exp /= 2;
-                base *= base;
-            }
-            exp -= 1;
-            result *= base;
-        }
-        result
+        Self(SMALL_FP.pow(self.0, SMALL_FP.from_elem(exp.0)))
     }
 
     /// Modular inverse
     ///
     /// Note: inverse of 0 is defined as 0.
     pub fn inv(self) -> Self {
-        // extended Euclidean
-        let mut x1: i32 = 1;
-        let mut a1: u32 = self.0;
-        let mut x0: i32 = 0;
-        let mut a2: u32 = MODULUS;
-        let mut q: u32 = 0;
-
-        while a2 != 0 {
-            let x2 = x0 - (q as i32) * x1;
-            x0 = x1;
-            let a0 = a1;
-            x1 = x2;
-            a1 = a2;
-            q = a0 / a1;
-            a2 = a0 - q * a1;
-        }
-        if x1 < 0 {
-            let (r, _) = MODULUS.overflowing_add(x1 as u32);
-            Field(r)
-        } else {
-            Field(x1 as u32)
-        }
+        Self(SMALL_FP.inv(self.0))
     }
 }
 
 impl From<u32> for Field {
     fn from(x: u32) -> Self {
-        Field(x % MODULUS)
+        Field(SMALL_FP.elem(x as u128))
     }
 }
 
 impl From<Field> for u32 {
     fn from(x: Field) -> Self {
-        x.0
+        SMALL_FP.from_elem(x.0) as u32
     }
 }
 
 impl PartialEq<u32> for Field {
     fn eq(&self, rhs: &u32) -> bool {
-        self.0 == *rhs
+        SMALL_FP.from_elem(self.0) as u32 == *rhs
```

Comment:
[Casting `u128` to `u32` will truncate](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=60c255ef47135400efb9a971b10f3cd5). I think you could instead promote `rhs` to `u128` to safely compare.

Reply:
Done.